pyfarm/agent/utility.py

```diff
@@ -275,3 +273,134 @@ def total_seconds(td):
+        assert isinstance(path, STRING_TYPES), path
+
+        try:
+            with open(path, "r") as loaded_file:
+                data = loaded_file.read().strip()
+
+            return UUID(data)
+
+        except (OSError, IOError) as e:
+            if e.errno != ENOENT:
+                raise
+
+        except ValueError:
+            cls.log.warning("%r does not contain valid data for a UUID", path)
+            return None
+
+    @classmethod
+    def _save(cls, agent_uuid, path):
+        """
+        Saves ``agent_uuid`` to ``path``.  This classmethod will
+        also create the necessary parent directories and handle
+        conversion from the input type :class:`uuid.UUID`.
+        """
+        assert isinstance(agent_uuid, UUID)
+        assert isinstance(path, STRING_TYPES)
+
+        parent_dir = dirname(path)
+        if parent_dir.strip():
+            try:
+                os.makedirs(parent_dir)
+
+            except (OSError, IOError) as e:
+                if e.errno != EEXIST:
+                    cls.log.warning("Failed to create %s, %s", parent_dir, e)
+                    raise
+
+        try:
+            with open(path, "w") as output:
+                output.write(str(agent_uuid))
+
+            cls.log.debug("Cached %s to %r", agent_uuid, path)
+            return path
+
+        except (OSError, IOError) as e:
+            cls.log.warning("Failed to write %s, %s", path, e)
+            raise
+
+    @classmethod
+    def load(cls, path=None):
+        """
+        This classmethod will attempt to load the agent's UUID from
+        disk.  If it cannot be located ``None`` will be returned.
+
+        :param string path:
+            An explicit path to load the UUID from.  In all other cases
+            we'll attempt to load the UUID from any directory
+            configuration files could load from.
+        """
+        cls.log.debug("Attempting to load agent UUID from disk")
+
+        if path is not None:
+            return cls._load(path)
+
+        for directory in config.directories(validate=False):
+            path = join(directory, "uuid.dat")
+            agent_uuid = cls._load(path)
+            if agent_uuid is not None:
+                cls.log.debug(
+                    "Cached UUID %s was loaded from %r", agent_uuid, path)
+                return agent_uuid
+
+        cls.log.warning(
+            "Unable to locate cached agent UUID, one will be generated.")
+
+    @classmethod
+    def save(cls, agent_uuid, path=None):
+        """
+        This classmethod will save ``agent_uuid`` to ``path`` or
+        to the first available path in the configuration.
+        """
+        assert isinstance(agent_uuid, UUID), agent_uuid
+        cls.log.debug("Attempting to save UUID %s to disk.", agent_uuid)
+
+        if path is not None:
+            return cls._save(agent_uuid, path=path)
+
+        directories = config.directories(validate=False, unversioned_only=True)
+        directories.reverse()
```

Comment:
There is still a problem with this that did not occur to me yesterday:

There is a very high likelyhood that this will end up choosing the current user's home directory. The system-wide configuration directory will likely not be writable, and the user's $HOME is next in the list.
The problem with this is, in large organisations, the home directories may well be network mounts or roaming profiles. (They are here.) This is a problem because then, different agent instances running on different machines but under the same user will have the same UUID.

We are going to have to rethink our strategy here. Going through ```config.directories()``` and saving the uuid in the first path where that works just isn't a good idea. There is no directory in there that is actually a good choice for that: The system-configuration directory is probably not writable, the user's home directory might be shared with other machines in some way and the current working directory is a bad choice in any case, because it could be anything.

My idea would be to have the full path to uuid.dat configured explicitly in the agent.yml config file (with some sensible defaults), and load the UUID and store it to this path, and this path only. Trying various different paths is just a potential source of confusion. (Example: An administrator wants to force a node to generate a new UUID for some reason. He finds one uuid.dat file and deletes it. For some reason, there is another uuid.dat in some other place that the agent still reads, so it retains its old uuid even though it isn't supposed to. If the administrator is lucky, they will at least notice that the planned UUID reset did not work.)
If we cannot access that path in the way we need to, just quit with an error message.

We might need different configuration keys for the differerent supported OS types (Linux, BSD, OS X, Windows), though.

Here's my idea for defaults:

- Linux, BSD: ```/etc/pyfarm-agent/uuid.dat``` (The user or the install script will have to make sure this directory exists and is writable by the pyfarm user.)
- Windows: ```%SYSTEMDRIVE%\pyfarm-agent\uuid.dat```
- OS X: No idea.